<a href="https://cognitiveclass.ai"><img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width = 400> </a>

<h1 align=center><font size = 5>Final Project : Regression Model in Keras</font></h1>


## Guide to reviewer


This notebook contains 4 parts A, B, C, and D which contain codes corresponding to the parts asked.


<h2>Regression Models with Keras</h2>

I'll be working upon the concrete data mentioned in the problem.

Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [1]:
import pandas as pd
import numpy as np

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 


#### Let's check how many data points we have.


In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


# PART : A

### 1. Building baseline model :
    1.1. 1 hidden layer with 10 nodes
    1.2. ReLU
    1.3. Adam
    1.4. MSE

### 2. Train Test Split.
### 3. Training in 50 epochs.
### 4. Evaluatin on test data, finding MSE
### 5. Repeating 2-4 50 times and creating list of 50 MSEs.
### 6. Finding meand and SD of MSEs.

In [10]:
import keras
from keras.models import Sequential
from keras.layers import Dense

# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [11]:
from sklearn.model_selection import train_test_split
# splitting data
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

# train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size = 0.2)

In [12]:
# predictors
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [13]:
# target
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [14]:
# number of predictor columns
n_cols = predictors.shape[1] # number of predictors

In [15]:
# Buidling, Fitting and Evaluating model
from sklearn.metrics import mean_squared_error
MSEs = []

# build the model
model = regression_model()

for index in range(50) :
    
    # fit the model
    res = model.fit(X_train, y_train, epochs=50, verbose=0)
    
    # predicting and calculating MSE
    y_pred = model.predict(X_test)
    MSEs.append(mean_squared_error(y_test, y_pred))

print(MSEs)    

7/7 [==============================] - 0s 370us/step
[146.67812076464614, 90.15180519551306, 60.64791303689099, 43.39411066410939, 39.706094853250605, 42.09299256968533, 37.933439451127136, 39.03671324833722, 38.52131726327032, 44.75912962391354, 37.53415731060051, 37.4512031800183, 38.02635567694607, 39.76753500789974, 37.12334832354841, 35.98090939136095, 38.289787965280716, 39.25955443513523, 38.147944578114426, 36.36263111102181, 37.554542644034434, 40.27078522901137, 39.51232577878876, 36.048989987580434, 36.02198599805725, 39.0065162906471, 37.84192895363892, 36.15224942116802, 41.13711928462836, 39.11387182902592, 37.23261406865879, 38.16613117595729, 38.13505828157632, 36.96632828498322, 37.309336112859654, 38.00534436805507, 38.838638558792404, 36.91192918365011, 35.845739549187385, 35.840163160552194, 37.25825091365575, 37.85355357337588, 35.921034569424606, 36.14434235329529, 35.04880048942609, 35.97336781974721, 41.06167187040737, 41.223763466942366, 39.68643670325594, 36.0

In [16]:
# finding mean and standard deviation of MSEs.
# importing statistics
import statistics

mean_mse = statistics.mean(MSEs)
stdev = statistics.stdev(MSEs)

print('Mean : %.2f' % mean_mse)
print('Standard Deviation : %.2f' % stdev)

Mean : 41.86
Standard Deviation : 17.20


# PART : B

## Part A with Normalized data

In [17]:
%%time
# Normalizing data
predictors_norm = (predictors - predictors.mean()) / predictors.std()
n_cols = predictors_norm.shape[1] # number of predictors

# train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size = 0.2)

# Buidling, Fitting and Evaluating model
from sklearn.metrics import mean_squared_error
MSEsB = []

# build the model
modelB = regression_model()

for index in range(50) :
    
    # fit the model
    modelB.fit(X_train, y_train, epochs=50, verbose=0)
    
    # predicting and calculating MSE
    y_pred = modelB.predict(X_test)
    MSEsB.append(mean_squared_error(y_test, y_pred))

    
print(MSEsB)

7/7 [==============================] - 0s 346us/step
[104.16897690907767, 54.50541586596538, 41.62055504151261, 37.91637974287171, 36.17117422239982, 34.71829639071134, 33.677447010110605, 32.419575630646875, 32.17116231824734, 31.693549233281804, 31.93192797682759, 31.62452602046483, 31.312140443278437, 31.11305357746709, 29.267788299014615, 28.941468516449884, 28.642527976689777, 28.416337506829894, 28.249193487496633, 27.683896684319357, 27.52055348628294, 28.135822577732053, 27.49231785116721, 27.19731945390886, 27.267392050410024, 26.422055546092018, 26.607335442932367, 26.465084044589698, 26.608118116164807, 26.72265731899872, 26.270020535087188, 25.937515579260946, 25.59744155766591, 26.10030082528517, 26.542560624804235, 25.57762971590291, 25.332321631716297, 25.661284492123265, 25.593150607550253, 24.96264100906396, 24.949107765311457, 25.996268889876124, 24.63765898054062, 24.610761037864645, 24.431265124514265, 24.51595814937806, 24.50087009549758, 25.11437103110908, 24.3683

In [18]:
# finding mean and standard deviation of MSEs.
# importing statistics
import statistics

mean_mseB = statistics.mean(MSEsB)
stdevB = statistics.stdev(MSEsB)

print('Mean : %.2f' % mean_mseB)
print('Standard Deviation : %.2f' % stdevB)

Mean : 30.25
Standard Deviation : 11.91


In [19]:
# Difference of means of A and B model
print('Difference of means of Model A and Model B is {}', abs(mean_mseB - mean_mse))

Difference of means of Model A and Model B is {} 11.60925141604362


# Part : C

### Part B with 100 epochs

In [20]:
%%time
# Normalizing data
predictors_norm = (predictors - predictors.mean()) / predictors.std()
n_cols = predictors_norm.shape[1] # number of predictors

# train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size = 0.2)

# Buidling, Fitting and Evaluating model
from sklearn.metrics import mean_squared_error
MSEsC = []

# build the model
modelC = regression_model()

for index in range(50) :
    
    # fit the model
    modelC.fit(X_train, y_train, epochs=100, verbose=0)
    
    # predicting and calculating MSE
    y_pred = modelC.predict(X_test)
    MSEsC.append(mean_squared_error(y_test, y_pred))

    
print(MSEsC)

7/7 [==============================] - 0s 332us/step
[111.10344380476121, 59.600438945123635, 42.25930167339531, 38.43099236306122, 38.09411563610241, 36.0305955354927, 35.671098576880055, 34.976045849487946, 34.37850874908017, 34.11387648492096, 33.502034134207854, 33.12362198489779, 32.6583801804382, 32.31419114916109, 31.357437653815342, 31.789054758250167, 30.969574807057064, 31.324510206024648, 30.967333868124804, 30.757438666591256, 30.601355104702368, 30.849444442694043, 30.816741508580588, 32.199493218185665, 30.851015386777124, 30.76409496888257, 30.659643669126343, 30.31168299132048, 29.876734909108865, 30.132353457686502, 29.651158438976378, 31.41158862490589, 29.726266522275633, 29.699157185859153, 30.277518770685056, 30.50852769581658, 29.99327823484942, 32.14665872954846, 29.74412548892853, 29.917347352557176, 29.974217115362496, 29.649065680290036, 29.98331239498231, 29.520980656786104, 30.543289339201294, 29.90268478050391, 30.126979710349946, 30.293559394185227, 29.076

In [21]:
# finding mean and standard deviation of MSEs.
# importing statistics
import statistics

mean_mseC = statistics.mean(MSEsC)
stdevC = statistics.stdev(MSEsC)

print('Mean : %.2f' % mean_mseC)
print('Standard Deviation : %.2f' % stdevC)

# Difference of means of A and B model
print('Difference of means of Model B and Model C is {}', abs(mean_mseB - mean_mseC))

Mean : 33.85
Standard Deviation : 12.11
Difference of means of Model B and Model C is {} 3.6034485330858814


# Part : D
### Part D is Part B with Three hidden layers, each of 10 nodes and ReLU activation function.

In [22]:
# define regression model
def regression_model_D():
    # create model
    modelD = Sequential()
    modelD.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    modelD.add(Dense(10, activation='relu')) # HL1
    modelD.add(Dense(10, activation='relu')) # HL2
    modelD.add(Dense(10, activation='relu')) # HL3
    modelD.add(Dense(1))
    
    # compile model
    modelD.compile(optimizer='adam', loss='mean_squared_error')
    return modelD

In [23]:
%%time
# Normalizing data
predictors_norm = (predictors - predictors.mean()) / predictors.std()
n_cols = predictors_norm.shape[1] # number of predictors

# train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size = 0.2)

# Buidling, Fitting and Evaluating model
from sklearn.metrics import mean_squared_error
MSEsD = []

# build the model
modelD = regression_model_D()

for index in range(50) :
    
    # fit the model
    modelD.fit(X_train, y_train, epochs=50, verbose=0)
    
    # predicting and calculating MSE
    y_pred = modelD.predict(X_test)
    MSEsD.append(mean_squared_error(y_test, y_pred))

    
print(MSEsD)

7/7 [==============================] - 0s 386us/step
[65.56721548150381, 39.39155738306632, 34.73529990465261, 32.86217002829231, 31.16024934844825, 30.98354730825358, 30.301381987708886, 29.365849212766673, 28.679988278895475, 28.96278385113849, 28.20485257422848, 28.366575771078555, 27.885929591880483, 28.079503040282653, 27.78585551240425, 28.707171231593996, 28.872783187801794, 28.271499623976776, 28.9646022169062, 28.299467114644976, 28.873607293916702, 29.307322335138426, 30.041080533136153, 29.145500970225992, 29.08503630901967, 28.668828394538213, 28.533114486704495, 29.504275597080017, 28.49390510481923, 27.909236311087994, 27.4093122566495, 28.122380059169124, 27.681079873048677, 27.312891420896488, 27.56198649964784, 27.621896015628646, 27.163841926538904, 27.105943126636987, 27.406187431590094, 27.85685434620073, 28.084843612240252, 27.728966904235936, 27.78505444922237, 28.022061428083028, 28.586845974151135, 28.715251606100086, 29.10992942337567, 28.70151923634619, 28.852

In [24]:
# finding mean and standard deviation of MSEs.
# importing statistics
import statistics

mean_mseD = statistics.mean(MSEsD)
stdevD = statistics.stdev(MSEsD)

print('Mean : %.2f' % mean_mseD)
print('Standard Deviation : %.2f' % stdevD)

# Difference of means of B and D model
print('Difference of means of Model B and Model D is {}', abs(mean_mseB - mean_mseD))

Mean : 29.68
Standard Deviation : 5.56
Difference of means of Model B and Model D is {} 0.5666340898626885


# End of Notebook